django/db/backends/mysql/operations.py

```diff
@@ -462,3 +464,16 @@ def on_conflict_suffix_sql(self, fields, on_conflict, update_fields, unique_fiel
             update_fields,
             unique_fields,
         )
+
+    def prepare_join_on_clause(self, lhs_table, lhs_field, rhs_table, rhs_field):
+        lhs_expr = Col(lhs_table, lhs_field)
+        rhs_expr = Col(rhs_table, rhs_field)
+
+        lhs_db_type = lhs_field.db_type(self.connection)
+        rhs_db_type = rhs_field.db_type(self.connection)
+        if lhs_db_type != rhs_db_type and not (
+            "integer" in lhs_db_type and "integer" in rhs_db_type
+        ):
+            rhs_expr = Cast(rhs_expr, lhs_field)
```

Comment:
This duplicates the `base` implement, but only the condition changes.
Should we refactor that one out to make it more DRY? 🤔 

What I could suggest `base/operations.py`:
```
def prepare_join_on_clause(...):
    ...
    if self.can_join_on_types(lhs_db_type, rhs_db_type):
    ...

def can_join_on_types(self, lhs_db_type, rhs_db_type):
    return lhs_db_type != rhs_db_type
```

and `mysql/operations.py`:
```
def can_join_on_types(self, lhs_db_type, rhs_db_type):
    return super().can_join_on_types(lhs_db_type, rhs_db_type) and not (
        "integer" in lhs_db_type and "integer" in rhs_db_type
    )
```

What do you think? :)